saver.py

```diff
@@ -181,5 +189,32 @@ def process_step(self, step):
         if step % self.config['save_steps'] == 0:
             self.save_model(f'step{step}')
 
+        pending_save_best_loss = os.path.exists(".pending_save_best_loss")
+        if pending_save_best_loss:
+            self.save_model('best_loss')
+            if is_main_process():
+                if self.old_best is not None:
+                    print(f'New best loss: {self.best_loss} from {self.old_best} (Δ{self.old_best - self.best_loss} = {100 * (self.old_best - self.best_loss) / self.old_best:.2f}%)')
+                else:
+                    print(f'New best loss: {self.best_loss}')
+                if os.path.exists(os.path.join(self.save_root, 'best_loss.txt')):
+                    os.remove(os.path.join(self.save_root, 'best_loss.txt'))
+                os.rename(".pending_save_best_loss", os.path.join(self.save_root, 'best_loss.txt'))
+
         if need_to_checkpoint(self.config):
             self.save_checkpoint(step)
+
+
+    def append_eval_results(self, loss, save_best=True):
+        if loss is not None:
+            if self.best_loss is None:
+                print(f"Evaluation loss: {loss}")
+            elif loss >= self.best_loss:
+                print(f"Evaluation loss: {loss} (best: {self.best_loss}, Δ: {self.best_loss - loss} = {100 * (self.best_loss - loss) / self.best_loss:.2f}%)")
+            if self.best_loss is None or loss < self.best_loss:
+                self.old_best = self.best_loss
+                self.best_loss = loss
+                if save_best:
+                    with open(".pending_save_best_loss", "w") as f:
```

Comment:
This file should probably be inside save_root. If you killed the job at the right time, it could persist before being renamed. Then if you launched another training run it might do weird things.

Reply:
Good point, moved.